common/reconciliation/invariant/staleWorkflow.go

```diff
@@ -0,0 +1,628 @@
+			zap.String("state", state), // may simplify log processing
+		)
+	}
+
+	if info.State == persistence.WorkflowStateRunning {
+		stale, expected, result := c.checkRunningAge(workflow, maxLifespan, domainName)
+		if stale {
+			// we know these exist, but we don't expect them to be numerous "recently".
+			// log for verification.
+			logStaleWorkflow("running", expected)
+		}
+		return stale, result
+	} else if info.State == persistence.WorkflowStateZombie {
+		// TODO: what exactly is zombie?  seems like probably "current but no concrete" but I'm surprised this is in the state.
+		//
+		// https://github.com/uber/cadence/issues/1800
+		// "The ability to create workflow with zombie status allows replication stack to backfill finished workflow execution sent by remote, while not affecting local running workflow"
+		// so these are just replication of completed workflows?  why is this a special state?
+		stale, expected, result := c.checkZombieAge(workflow, maxLifespan, domainName)
+		if stale {
+			// we know these exist, but we don't have a good feel for the distribution
+			logStaleWorkflow("zombie", expected)
+		}
+		return stale, result
+	}
+	stale, expected, result := c.checkClosedAge(workflow, maxLifespan, domainName)
+	if stale {
+		// we know these exist, but we don't have a good feel for the distribution
+		logStaleWorkflow("closed", expected)
+	}
+	return stale, result
+}
+
+func (c *staleWorkflowCheck) checkRunningAge(workflow *persistence.GetWorkflowExecutionResponse, maxLifespan time.Duration, domainName string) (pastExpiration bool, expected time.Time, result CheckResult) {
+	// workflow-expiration timer is calculated in GenerateWorkflowStartTasks, mimic it here for safety: https://github.com/uber/cadence/blob/master/service/history/execution/mutable_state_task_generator.go#L140-L163
+	/*
+		// running workflows might contain critical information in their first history record, so it must be retrieved.
+		info := workflow.State.ExecutionInfo
+
+		duration := time.Duration(info.WorkflowTimeout) * time.Second
+		// First-decision-backoff is only in first event.  It must be included because the expiration time does not include it.
+		// I really wish it wasn't built this way, this makes it much much harder to scan in bulk.
+		timeout := info.StartTimestamp.Add(duration + firstdecisionbackoff)
+
+		// ensure that the first attempt does not time out early based on retry policy timeout
+		if attr.Attempt > 0 && !executionInfo.ExpirationTime.IsZero() && workflowTimeoutTimestamp.After(executionInfo.ExpirationTime) {
+			workflowTimeoutTimestamp = executionInfo.ExpirationTime
+		}
+	*/
+	first, ok, result := c.firstEvent(workflow, domainName)
+	if !ok {
+		// completely missing history should be caught by the history-exists check, no need to handle here
+		return false, expected, result
+	}
+	startedAt := workflow.State.ExecutionInfo.StartTimestamp                             // nonzero
```

Comment:
I didn't catch this earlier, but perhaps add your impossibly-early start guard here in-case there was some accidental nil start values?

Reply:
if startedAt.Before(impossiblyOld) {
    return false, ....
}

Reply:
oh, you're doing it below, nevermind